Client-3517: Dynamically apply dynamic.client stanza of config

paul-brackin opened the pull request.

Commit:
update cluster and node members as appropriate from dynamic client config

Commit:
use local CP var, use tmp array, validate max_socket_idle upstream

mirzakaracic commented on client/src/com/aerospike/client/cluster/Cluster.java:
```diff
@@ -397,6 +397,46 @@ else if (policy.user != null && policy.user.length() > 0) {
+
+		if (clientPolicy.maxSocketIdle == 0) {
+			maxSocketIdleNanosTran = 0;
+			maxSocketIdleNanosTrim = TimeUnit.SECONDS.toNanos(55);
```
issue: You have magic number here

mirzakaracic commented on client/src/com/aerospike/client/cluster/Cluster.java:
```diff
@@ -397,6 +397,46 @@ else if (policy.user != null && policy.user.length() > 0) {
+				rackIdsTemp[i] = list.get(i);
+			}
+		}
+		else {
```
nit: if else is not formatted as other code

mirzakaracic commented on client/src/com/aerospike/client/policy/ClientPolicy.java:
```diff
@@ -463,7 +465,11 @@ public ClientPolicy(ClientPolicy other, ConfigurationProvider configProvider) {
 		if (dynCC.maxSocketIdle != null) {
-			this.maxSocketIdle = dynCC.maxSocketIdle.value;
+			if (dynCC.maxSocketIdle.value < 0) {
+				Log.error("Invalid maxSocketIdle in config: " + dynCC.maxSocketIdle.value);
```
nit: it is good that we are checking for valid types but we are not validating types for anything else.

mirzakaracic reviewed (commented):
Couple minor things. Looks great otherwise.

Commit:
refactor / commonize the application of the dynamic config properties, create constant variable definition for MAX_SOCKET_IDLE_TRIM_DEFAULT_SECS

mirzakaracic reviewed (dismissed):
👍

Commit:
check for invalid maxSocketIdle for the case of default client policy passed in

Commit:
move const to top of class, check temp racks array before copying & doing node rack ops, remove unused import

BrianNichols committed:
Only change rackIds and "node.racks" if the new rackIds are different.
Make an reference copy in getRackNode().

Commit:
refactor YamlConfigProvider for efficiency and more consistent ex handling.  fix up associated tests

BrianNichols committed:
Cleanup YamlConfigProvider.
Add YamlConfigProvider.getConfigProvider().
Move CONFIG_PATH_ENV and CONFIG_PATH_SYS_PROP to YamlConfigProvider.

Commit:
fix tests to match new ex msg

Commit:
remove stackdumps from log

Commit:
alter tend thread hook conditions.  set default config interval time

Commit:
fix default configInterval time

BrianNichols committed:
Add new configPath variable. configPath is retrieved from environment variable(s).
If configPath is not null, dynamic config loads are always attempted.
If configPath is null, dynamic config is permanently disabled.

BrianNichols committed:
Validate tendInterval after every dynamic config load.
Make Cluster configInterval a duration (default 60000).
Convert config interval from a duration to the number of cluster tend iterations just before checking the dynamic config file timestamp.

Commit:
update yaml config file
